### PR TITLE
Better analysis pipeline

### LINKED ISSUE #292: Implement the Display trait for the assembly ast.

### PR DESCRIPTION
**Pull request by Schaeff**
parse -> expand_macros -> analyse -> pilgen -> parse -> expand_macros -> pil_analyze

How:
- I put all the ast in a new `ast` crate (except pil_analyzer which still uses its own ast and I think it's fine for now)
- I put expanding macros and analysis (currently the simple label batching) in a new `analysis` crate

Impact:
- asm blocks don't work anymore
- that's it

fixes #292

### CODE REVIEW

File: compiler/tests/pil.rs [hunk 1]
```diff
@@ -76,31 +76,3 @@ fn test_pair_lookup() {
 fn test_block_lookup_or() {
     verify_pil("block_lookup_or.pil", None);
 }
-
-#[test]
-fn test_simple_sum_asm_pil() {
```
Comment (Schaeff) [inline compiler/tests/pil.rs:81]: Removed these as they contain assembly blocks

File: ast/src/analysis/mod.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+use num_bigint::BigUint;
+
+use crate::parsed::{
+    asm::{InstructionBodyElement, InstructionParams, RegisterFlag},
+    Expression, Statement,
+};
+
+pub struct RegisterDeclarationStatement {
+    pub start: usize,
```
Comment (leonardoalt) [inline ast/src/analysis/mod.rs:13]: What's start? (here and in other structs)

File: ast/src/parsed/display.rs [hunk 3]
```diff
@@ -195,40 +181,8 @@ impl Display for InstructionParam {
...
-            "{}",
-            self.0
-                .iter()
-                .map(|r| r.to_string())
-                .collect::<Vec<_>>()
-                .join(", ")
-        )
-    }
+pub fn quote(input: &str) -> String {
+    format!("\"{}\"", input.replace('\\', "\\\\").replace('"', "\\\""))
```
Comment (leonardoalt) [inline ast/src/parsed/display.rs:185]: Wasn't this added in a different PR?

File: ast/src/parsed/display.rs [hunk 4]
```diff
@@ -101,6 +55,38 @@ impl<T: Display> Display for ASMStatement<T> {
...
+                write!(
+                    f,
+                    "pil{{\n{}\n}}",
+                    statements
+                        .iter()
+                        .map(|s| format!("{}", s))
+                        .collect::<Vec<_>>()
+                        .join("\n")
+                )
+            }
```
Comment (leonardoalt) [inline ast/src/parsed/display.rs:89]: Why were these moved around? Looks like almost the same content?

File: ast/src/analysis/mod.rs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    pub name: String,
+}
+
+pub struct PilBlock<T> {
+    pub start: usize,
+    pub statements: Vec<Statement<T>>,
+}
+
+pub struct AnalysisASMFile<T> {
```
Comment (leonardoalt) [inline ast/src/analysis/mod.rs:76]: Nice!

File: ast/src/analysis/mod.rs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+#[derive(Default, Debug, PartialEq, Eq, Clone)]
+pub struct BatchMetadata {
+    // the set of compatible statements
+    pub size: usize,
+    // the reason why this batch ended (for debugging purposes), None if we ran out of statements to batch
+    pub reason: Option<IncompatibleSet>,
+}
+
+impl BatchMetadata {
+    pub fn get_size(&self) -> usize {
```
Comment (leonardoalt) [inline ast/src/analysis/mod.rs:94]:
```suggestion
    pub fn size(&self) -> usize {
```
personal taste, usually I don't like `get*` named functions

**Review by leonardoalt — COMMENTED** [review]
Review for the new `ast` crate.

File: ast/src/analysis/mod.rs (hunk 2 above)
Comment (Schaeff) [reply ast/src/analysis/mod.rs:13]: I think it's span information from the parser

File: ast/src/parsed/display.rs (hunk 4 above)
Comment (Schaeff) [reply ast/src/parsed/display.rs:89]: good call will fix

File: ast/src/parsed/display.rs (hunk 3 above)
Comment (Schaeff) [reply ast/src/parsed/display.rs:185]: I moved it here from `parser_utils` as it's only used here

File: analysis/src/batcher.rs [hunk 7]
```diff
@@ -1,19 +1,22 @@
...
-use parser::asm_ast::{
-    batched::{ASMStatementBatch, BatchedASMFile, Incompatible, IncompatibleSet},
-    ASMFile, ASMStatement,
-};
+
+pub fn batch<T: FieldElement>(file: AnalysisASMFile<T>) -> AnalysisASMFile<T> {
+    ProgramBatcher::default().batch(file)
+}
 
 #[derive(Default)]
```
Comment (Schaeff) [inline analysis/src/batcher.rs:13]: I refactored the batcher to work with the new analysis structure: now batches as stored separately from statements

File: ast/src/parsed/display.rs (hunk 4 above)
Comment (leonardoalt) [reply ast/src/parsed/display.rs:61]: fine as is, just wanted to know if there was more to it

Comment (Schaeff) [reply ast/src/parsed/display.rs:61]: There isnt, I changed it back the diff should be smaller

File: analysis/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+[dependencies]
+ast = { path = "../ast" }
+itertools = "0.10.5"
+log = "0.4.18"
+number = { version = "0.1.0", path = "../number" }
+pretty_assertions = "1.3.0"
+
+[dev-dependencies]
+parser = { path = "../parser" }
```
Comment (leonardoalt) [inline analysis/Cargo.toml:16]: Where is this used?

File: analysis/src/type_check.rs [hunk 9]
```diff
@@ -0,0 +1,74 @@
+use ast::{
+    analysis::{
+        AnalysisASMFile, AssignmentStatement, DegreeStatement, InstructionDefinitionStatement,
+        InstructionStatement, LabelStatement, PilBlock, RegisterDeclarationStatement,
+    },
+    parsed::asm::{ASMFile, ASMStatement},
+};
+use number::FieldElement;
+
+/// A very stupid type checker. TODO: make it smart
```
Comment (leonardoalt) [inline analysis/src/type_check.rs:10]: Can you make an issue for it so we remember?

File: analysis/src/macro_expansion.rs [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+use ast::parsed::{
+    asm::{ASMFile, ASMStatement, InstructionBodyElement},
+    postvisit_expression_mut, ArrayExpression, Expression, FunctionDefinition, SelectedExpressions,
+    Statement,
+};
+use number::FieldElement;
+
+pub fn expand<T: FieldElement>(file: ASMFile<T>) -> ASMFile<T> {
+    let mut expander = MacroExpander::default();
+    expander.expand_asm(file)
```
Comment (leonardoalt) [inline analysis/src/macro_expansion.rs:12]: Does this macro expander still use the asm/pil context outside of the macro definition?

Comment (leonardoalt) [reply analysis/src/macro_expansion.rs:12]: Ah, I guess this is the same as the other file and was just moved to the new crate.

Comment (leonardoalt) [reply analysis/src/macro_expansion.rs:12]: Hm it does look different. Can you explain what the differences are?

File: compiler/src/verify.rs [hunk 11]
```diff
@@ -16,7 +18,13 @@ pub fn compile_asm_string_temp<T: FieldElement>(
...
     inputs: Vec<T>,
 ) -> (String, Temp) {
-    let pil = pilgen::compile(Some(file_name), contents).unwrap();
+    let parsed = parser::parse_asm(Some(file_name), contents).unwrap_or_else(|err| {
+        eprintln!("Error parsing .asm file:");
+        err.output_to_stderr();
+        panic!();
+    });
+    let analysed = analyse(parsed).unwrap();
+    let pil = convert(analysed);
```
Comment (leonardoalt) [inline compiler/src/verify.rs:27]: Maybe `convert` could do the parsed and analysed parts inside itself? Or did you keep it separated on purpose?

File: analysis/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+mod macro_expansion;
+mod type_check;
+
+/// expose the macro expander for use in the pil_analyzer
+pub use macro_expansion::MacroExpander;
+
+use ast::{analysis::AnalysisASMFile, parsed::asm::ASMFile};
+use number::FieldElement;
+
+pub fn analyse<T: FieldElement>(file: ASMFile<T>) -> Result<AnalysisASMFile<T>, String> {
```
Comment (leonardoalt) [inline analysis/src/lib.rs:11]:
```suggestion
pub fn analyze<T: FieldElement>(file: ASMFile<T>) -> Result<AnalysisASMFile<T>, String> {
```

**Review by leonardoalt — COMMENTED** [review]
Missing review for pil_analyzer and pilgen.

File: pilgen/src/lib.rs [hunk 13]
```diff
@@ -1,47 +1,24 @@
...
-) -> Vec<Statement<T>> {
-    let batched_asm = ASMBatcher::default().convert(ASMFile(statements.into_iter().collect()));
-    ASMPILConverter::new(macro_expander).convert(batched_asm, ASMKind::Inline)
-}
-
-#[derive(PartialEq)]
-pub enum ASMKind {
-    Inline,
-    StandAlone,
+pub fn convert<T: FieldElement>(analysis: AnalysisASMFile<T>) -> PILFile<T> {
```
Comment (leonardoalt) [inline pilgen/src/lib.rs:20]: Why `convert` and not `compile`?

File: pilgen/src/lib.rs [hunk 14]
```diff
@@ -1054,7 +1029,9 @@ pol constant p_reg_write_X_CNT = [1, 0, 0, 0, 0, 0, 0, 0] + [0]*;
 "#;
         let file_name = "../test_data/asm/simple_sum.asm";
         let contents = fs::read_to_string(file_name).unwrap();
-        let pil = compile::<GoldilocksField>(Some(file_name), &contents).unwrap();
+        let parsed = parse_asm::<GoldilocksField>(Some(file_name), &contents).unwrap();
+        let analysis = analyse(parsed).unwrap();
+        let pil = convert(analysis);
```
Comment (leonardoalt) [inline pilgen/src/lib.rs:1034]: A helper for these 3 calls could be useful.

**Review by leonardoalt — COMMENTED** [review]
I'm done reviewing everything for now. Will wait for the comments to be addressed to look into it again.

File: analysis/Cargo.toml (hunk 8 above)
Comment (Schaeff) [reply analysis/Cargo.toml:16]: In the tests which compare blocks of texts, this gives you a diff. Will move to dev-deps!

File: analysis/src/type_check.rs (hunk 9 above)
Comment (Schaeff) [reply analysis/src/type_check.rs:10]: #338

File: analysis/src/lib.rs (hunk 12 above)
Comment (Schaeff) [reply analysis/src/lib.rs:11]: 🇬🇧

File: analysis/src/macro_expansion.rs (hunk 10 above)
Comment (Schaeff) [reply analysis/src/macro_expansion.rs:15]: Refactored this based on the current version on `main`, the diff is much smaller now

File: analysis/src/lib.rs (hunk 12 above)
Comment (Schaeff) [reply analysis/src/lib.rs]: ok fine

**Review by leonardoalt — APPROVED** [review]
